### PR TITLE
[6581] Funding CSV Upload

### PR DESCRIPTION
**Pull request by AdColvin**
https://trello.com/c/wJJuyo0V

### Context
This adds a job called ProcessFundingCsvJob, which calls a new FundingDataImporter service. This service determines what type of import is needed and uses the services classes currently being used in [funding.rake](https://github.com/DFE-Digital/register-trainee-teachers/blob/6580-funding-csv-upload/lib/tasks/funding.rake) 

I'm basing the type of import on the name of the file, based on the instructions given [here](https://github.com/DFE-Digital/register-trainee-teachers/blob/6580-funding-csv-upload/docs/uploading_funding_csvs.md). But if there's a better way than this then do let me know. 

### Important business

- [ ] Does this PR introduce any PII fields that need to be overwritten or deleted in db/scripts/sanitise.sql?
- [ ] Does this PR change the database schema? If so, have you updated the config/analytics.yml file and considered whether you need to send 'import_entity' events?
- [ ] Do we need to send any updates to DQT as part of the work in this PR?
- [ ] Does this PR need an ADR?

NB: Please notify the #twd_data_insights team and ask for a review if new fields are being added to analytics.yml

### CODE REVIEW

File: app/jobs/process_funding_csv_job.rb [hunk 1]
```diff
@@ -0,0 +1,8 @@
+class ProcessFundingCsvJob < ApplicationJob
```
Comment (darokel) [inline app/jobs/process_funding_csv_job.rb:1]: I think this job will also need to be added somewhere here https://github.com/DFE-Digital/register-trainee-teachers/pull/3930/files#diff-b03c967c04be863ba7c78e181a3bf217f068b9bac69ccef018a9ebbee5bf3c80R19 right after the form is submitted and the `FundingUpload` record is created.

**Comment by AdColvin** [comment]
@darokel I've made the changes you suggested, I'm now passing a FundingUpload record to the job and switching on the funding_type to decide which service to use. I had to update the parser classes and specs as well to allow for that. 

I'm tracking the status of the upload from the FundingDataImporter, setting it to pending, processed or failed at the appropriate time.  

I'm also now queuing the job from the FundingUploadsController if the save is successful. 

Let me know what you think! Cheers

**Comment by darokel** [comment]
@AdColvin I think this needs a rebase then I'll take another look 👀

File: app/lib/funding/parsers/base.rb [hunk 2]
```diff
@@ -12,8 +12,8 @@ def expected_headers
           raise(NotImplementedError("implement in subclass"))
         end
 
-        def to_attributes(file_path:)
-          csv = CSV.open(file_path, headers: true)
+        def to_attributes(funding_upload)
+          csv = CSV.parse(funding_upload.csv_data, headers: true)
```
Comment (darokel) [inline app/lib/funding/parsers/base.rb:16]: I think we still need to support a regular csv file path for the rake task incase it needs to be done manually. We could check for either and work off that?

**Review by darokel — COMMENTED** [review]
This all looks good to me. I left one last comment about whether we still want to support the csv path and not completely replace it with a funding upload record to still use the rake task when needed

**Comment by AdColvin** [comment]
@darokel I've updated the base parser now to accept the same `file_path` argument that it did before. So the rake task can still be used. Let me know what you think 👍

**Review by darokel — APPROVED** [review]
Nice work!

**Pull request merged**